src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,281 @@
+
+	ColumnarMetapage metapage;
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *) &metapage, sizeof(ColumnarMetapage));
+
+	uint64 stripe_id = metapage.reserved_stripe_id;
+	metapage.reserved_stripe_id++;
+
+	*first_row = metapage.reserved_row_number;
+	metapage.reserved_row_number += nrows;
+
+	WriteToBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				 (char *) &metapage, sizeof(ColumnarMetapage), true);
+
+	UnlockRelationForExtension(rel, ExclusiveLock);
+
+	return stripe_id;
+}
+
+void
+ColumnarStorageRead(Relation rel, uint64 logicalOffset, char *data, uint32 amount)
+{
+	uint64 read = 0;
+
+	Assert(LogicalOffsetIsValid(logicalOffset));
+
+	while (read < amount)
+	{
+		SmgrAddr addr = logical_to_smgr(logicalOffset + read);
+
+		uint32 to_read = Min(amount - read, BLCKSZ - addr.offset);
+		ReadFromBlock(rel, addr.blockno, addr.offset, data + read, to_read);
+
+		read += to_read;
+	}
+}
+
+
+void
+ColumnarStorageWrite(Relation rel, uint64 logicalOffset, char *data, uint32 amount)
+{
+	Assert(LogicalOffsetIsValid(logicalOffset));
+
+	while (amount > 0)
+	{
+		SmgrAddr addr = logical_to_smgr(logicalOffset);
+
+		uint64 to_write = Min(BLCKSZ - addr.offset, amount);
+		WriteToBlock(rel, addr.blockno, addr.offset, data, to_write, false);
+
+		data += to_write;
+		amount -= to_write;
+		logicalOffset += to_write;
+	}
+}
+
+
+/*
+ * ReadFromBlock - read bytes from a page at the given offset.
+ */
+static void
+ReadFromBlock(Relation rel, BlockNumber blockno, uint32 offset, char *buf,
+			  uint32 len)
+{
+	Buffer buffer = ReadBuffer(rel, blockno);
+	Page page = BufferGetPage(buffer);
+	PageHeader phdr = (PageHeader) page;
+
+	Assert(offset + len <= phdr->pd_upper);
+	memcpy_s(buf, phdr->pd_upper - offset, page + offset, len);
+	ReleaseBuffer(buffer);
+}
+
+/*
+ * WriteToBlock - append data to a block, initializing if necessary, and emit
+ * WAL. If 'clear' is true, always clear the data on the page and reinitialize
+ * it first, and offset must be SizeOfPageHeaderData.
+ *
+ * The data must fit on the page, and pd_lower will be set to the end of the
+ * written data.
+ */
+static void
+WriteToBlock(Relation rel, BlockNumber blockno, uint32 offset, char *buf,
+			 uint32 len, bool clear)
+{
+	Buffer buffer = ReadBuffer(rel, blockno);
+	LockBuffer(buffer, BUFFER_LOCK_EXCLUSIVE);
+
+	Page page = BufferGetPage(buffer);
+	PageHeader phdr = (PageHeader) page;
+	if (PageIsNew(page) || clear)
+	{
+		PageInit(page, BLCKSZ, 0);
+	}
+
+	/*
+	 * Ordinarily, the offset should be pd_lower, because we are
+	 * appending. But after a transaction has been rolled-back, we might be
+	 * over-writing the rolled-back data (if it's at the end), so
+	 * phdr->pd_lower can be higher. In that case, we emit a lot message and
+	 * reset pd_lower to offset.
+	 */
+	if (phdr->pd_lower > offset)
+	{
+		ereport(DEBUG1, (errmsg("over-writing page %u", blockno),
```

Comment:
DEBUG1 sounds little bit high to me, what about DEBUG2 ?
Also, would be nice to have a test for that (if possible)

Reply:
(resolving that since we don't have this line anymore)